moveit_ros/planning/planning_request_adapter_plugins/src/fix_start_state_collision.cpp

```diff
@@ -115,37 +126,86 @@ class FixStartStateCollision : public planning_request_adapter::PlanningRequestA
+      for (std::size_t i = 0; i < jmodels.size(); ++i)
+        link_models[i] = jmodels[i]->getChildLinkModel();
+
+      // Also record end effector link pose if one is set
+      if (!req.goal_constraints[0].position_constraints.empty())
+        link_models.push_back(start_state.getLinkModel(req.goal_constraints[0].position_constraints[0].link_name));
+      else if (!req.goal_constraints[0].orientation_constraints.empty())
+        link_models.push_back(start_state.getLinkModel(req.goal_constraints[0].orientation_constraints[0].link_name));
+
+      EigenSTL::vector_Isometry3d original_link_poses;
+      original_link_poses.resize(link_models.size());
+      Eigen::Isometry3d new_link_pose;
+
+      for (std::size_t i = 0; i < link_models.size(); ++i)
+        original_link_poses[i] = start_state.getGlobalLinkTransform(link_models[i]);
+
+      double dist, max_dist;
+      double local_jiggle_fraction_ = jiggle_fraction_;
       bool found = false;
       for (int c = 0; !found && c < sampling_attempts_; ++c)
       {
+        // Change each joint slightly ("jiggle")
         for (std::size_t i = 0; !found && i < jmodels.size(); ++i)
         {
           std::vector<double> sampled_variable_values(jmodels[i]->getVariableCount());
-          const double* original_values = prefix_state->getJointPositions(jmodels[i]);
+          const double* original_values = original_state->getJointPositions(jmodels[i]);
           jmodels[i]->getVariableRandomPositionsNearBy(rng, &sampled_variable_values[0], original_values,
                                                        jmodels[i]->getMaximumExtent() * jiggle_fraction_);
           start_state.setJointPositions(jmodels[i], sampled_variable_values);
-          collision_detection::CollisionResult cres;
-          planning_scene->checkCollision(creq, cres, start_state);
-          if (!cres.collision)
+        }
+
+        // Confirm that no links move far after jiggling
+        max_dist = -1;
+        bool euclidean_distance_ok = true;
+        for (std::size_t i = 0; !found && i < link_models.size(); ++i)
+        {
+          new_link_pose = start_state.getGlobalLinkTransform(link_models[i]);
+          dist = pow(new_link_pose.translation().x() - original_link_poses[i].translation().x(), 2) +
+                 pow(new_link_pose.translation().y() - original_link_poses[i].translation().y(), 2) +
+                 pow(new_link_pose.translation().z() - original_link_poses[i].translation().z(), 2);
+
+          if (dist > max_dist)
+            max_dist = dist;
+          if (dist > max_jiggle_distance_squared_)
           {
-            found = true;
-            ROS_INFO("Found a valid state near the start state at distance %lf after %d attempts",
-                     prefix_state->distance(start_state), c);
+            // Reduce jiggle fraction if it caused excessive movement
+            local_jiggle_fraction_ = local_jiggle_fraction_ * ((sampling_attempts_ - 2) / sampling_attempts_);
```

Comment:
This breaks if `sampling_attempts_ <= 2`.

I'm not sold on the simulated annealing. Did you weigh up against alternatives, e.g. no annealing, local_jiggle_fraction's for each joint, stochastic sample projection to within the maximum allowed distance?

Alternatively I could imaging changing to sample a joint diff vector (or a number of them) for directions and add this vector in a loop until we exceed the Cartesian limit, checking each intermediate joint.
Of course the best solution would be an informed procedure using Jacobian and penetration depth information, but I think that would end up with different problems in terms of complexity.

Reply:
True, I blindly assumed sampling_attempts (that's actually the max attempts, not the counter) would not be that low.

No, I didn't do any quantitative investigation, I just added it like that because I figured it couldn't hurt. I don't think it's worth the effort to look into it that deeply (this is an adapter for edge cases). It would be better spent on using penetration depth and making this work in Cartesian space as you described.